### PR TITLE
Update to serde 0.9.X

### LINKED ISSUE #92: Support serde 0.9.X
Serde 0.9.0-rc1 is out (https://github.com/serde-rs/serde/issues/668) and there are some breaking changes.
It would be great to have a support for it. Thanks.

### PR DESCRIPTION
**Pull request by dgriffen**
Closes #92 

One of the few questions I had left after doing this was what to do with the `isize_invalid_deserialize` test. The test used variants from the `serde::de::value::Error` enum. This type is now a struct and doesn't have variants anymore so the test wouldn't work without string parsing, which seemed fragile.

### CODE REVIEW

File: src/serde/reader.rs [hunk 1]
```diff
@@ -303,162 +275,193 @@ impl<'a, R: Read> serde::Deserializer for Deserializer<'a, R> {
...
+            fn visit_variant_seed<V>(self, seed: V) -> DeserializeResult<(V::Value, Self::Variant)>
+                where V: serde::de::DeserializeSeed,
+            {
+                let idx: u32 = try!(serde::de::Deserialize::deserialize(&mut *self.0));
+                let val: Result<_, DeserializeError> = seed.deserialize(idx.into_deserializer());
+                Ok((try!(val), self))
+            }
+        }
+
+        impl<'a, R: Read + 'a> serde::de::VariantVisitor for VariantVisitor<'a, R> {
```
Comment (dtolnay) [inline src/serde/reader.rs]: We already have a VariantVisitor below don't we? Let's use that one.

```rust
impl<'a, R: Read> serde::de::VariantVisitor for &'a mut Deserializer<R> { /* ... */ }
```

File: src/serde/reader.rs [hunk 2]
```diff
@@ -303,162 +275,193 @@ impl<'a, R: Read> serde::Deserializer for Deserializer<'a, R> {
...
+    fn deserialize_enum<V>(self,
                      _enum: &'static str,
                      _variants: &'static [&'static str],
                      mut visitor: V) -> Result<V::Value, Self::Error>
-        where V: serde::de::EnumVisitor,
+        where V: serde::de::Visitor,
     {
-        visitor.visit(self)
-    }
+        struct VariantVisitor<'a, R: Read + 'a>(&'a mut Deserializer<R>);
```
Comment (dtolnay) [inline src/serde/reader.rs]: I think this would be clearer without the wrapper type:

```rust
impl<'a, R: Read> serde::de::EnumVisitor for &'a mut Deserializer<R>
```

File: src/serde/reader.rs [hunk 3]
```diff
@@ -97,12 +97,8 @@ impl fmt::Display for DeserializeError {
 }
 
 impl serde::de::Error for DeserializeError {
-    fn custom<T: Into<String>>(desc: T) -> DeserializeError {
-        DeserializeError::Serde(serde::de::value::Error::Custom(desc.into()))
-    }
-
-    fn end_of_stream() -> DeserializeError {
-        DeserializeError::Serde(serde::de::value::Error::EndOfStream)
+    fn custom<T: ::std::fmt::Display>(desc: T) -> DeserializeError {
```
Comment (dtolnay) [inline src/serde/reader.rs]: The `fmt` module is already imported above so this could be `fmt::Display` instead of `::std::fmt::Display`.

File: src/serde/writer.rs [hunk 4]
```diff
@@ -29,17 +29,17 @@ pub enum SerializeError {
...
 }
 
 fn wrap_io(err: IoError) -> SerializeError {
     SerializeError::IoError(err)
 }
 
 impl serde::ser::Error for SerializeError {
-    fn custom<T: Into<String>>(msg: T) -> Self {
-        SerializeError::Custom(msg.into())
+    fn custom<T: ::std::fmt::Display>(msg: T) -> Self {
```
Comment (dtolnay) [inline src/serde/writer.rs]: Please use `fmt::Display` instead of `::std::fmt::Display`.

File: src/serde/writer.rs [hunk 5]
```diff
@@ -29,17 +29,17 @@ pub enum SerializeError {
...
+        SerializeError::Custom(format!("{}", msg))
```
Comment (dtolnay) [inline src/serde/writer.rs]: I think this would be clearer as `msg.to_string()` instead of `format!("{}", msg)`.

Comment (dgriffen) [inline src/serde/writer.rs]: Should I make these updates as separate commits? or squash all the changes into one commit?

**Comment by dtolnay** [comment]
Regarding the error tests - I am not sure why `DeserializeError::Serde` contains `serde::de::value::Error`. I think it should contain String instead, just like the `SerializeError`. And the tests should do string matching - this is what serde_json does.

File: src/serde/writer.rs (hunk 5 above)
Comment (dtolnay) [inline src/serde/writer.rs]: Please make any changes as new commits. GitHub's code review has very poor support for reviewing squashed / amended / rebased commits so I would have to review the whole thing all over again to see what you changed. If he prefers, @TyOverby can squash or rebase the whole thing at the end when the review is done and it is ready to merge.

**Comment by dgriffen** [comment]
Should I re-add the test and change `DeserializeError::Serde` to contain a string then?

**Comment by dtolnay** [comment]
Yes please change the error to a String and update the deserializing_errors test to check the error messages.

File: src/serde/reader.rs [hunk 6]
```diff
@@ -54,7 +54,7 @@ impl Error for DeserializeError {
             DeserializeError::IoError(ref err) => Error::description(err),
             DeserializeError::InvalidEncoding(ref ib) => ib.desc,
             DeserializeError::SizeLimit => "the size limit for decoding has been reached",
-            DeserializeError::Serde(ref s) => s.description(),
+            DeserializeError::Custom(_) => "a custom serialization error was reported",
```
Comment (dtolnay) [inline src/serde/reader.rs]: You should be able to return the real error here. `DeserializeError::Custom(ref msg) => msg`

File: src/serde/reader.rs [hunk 7]
```diff
@@ -77,7 +77,7 @@ impl From<IoError> for DeserializeError {
 
 impl From<serde::de::value::Error> for DeserializeError {
```
Comment (dtolnay) [inline src/serde/reader.rs]: We don't need this impl anymore.

File: src/serde/reader.rs [hunk 8]
```diff
@@ -90,19 +90,15 @@ impl fmt::Display for DeserializeError {
                 write!(fmt, "InvalidEncoding: {}", ib),
             DeserializeError::SizeLimit =>
                 write!(fmt, "SizeLimit"),
-            DeserializeError::Serde(ref s) =>
-                s.fmt(fmt),
+            DeserializeError::Custom(ref s) =>
+                write!(fmt, "Custom Error {}", s),
```
Comment (dtolnay) [inline src/serde/reader.rs]: I would prefer to just format the message like before - `s.fmt(fmt)`. I don't think "Custom Error" adds any clarity or context.

Comment (dgriffen) [inline src/serde/reader.rs]: Sounds good. I was just following the convention used by SerializeError

Comment (dtolnay) [inline src/serde/reader.rs]: A major focus of serde 0.9 is better error message so as the person doing this upgrade, part of your job is improving existing bad error messages!

**Comment by dtolnay** [comment]
I tried compiling this and I get a ton of unused_import, unused_mut, and unused_variables warnings. The master branch builds cleanly so please fix all of the warnings.

**Comment by dgriffen** [comment]
I wasn't aware those errors were introduced with my changes. I'll get those corrected.

**Comment by TyOverby** [comment]
This is really great work @dgriffen, thanks for taking this PR on!

From the bincode API side, this all looks good to me.  @dtolnay: is there anything remaining as far as integration with serde is concerned?

**Comment by TyOverby** [comment]
🎈 🎈

**Comment by dgriffen** [comment]
@TyOverby are there any more blocking issues that a new release is waiting on? I'd be willing to work on them.

**Comment by dtolnay** [comment]
Possibly https://github.com/TyOverby/bincode/issues/86.

**Comment by TyOverby** [comment]
@dgriffen: #86 and #94 were the next ones that I was going to take on. 

I also just opened up #95 which is a precondition for #86.

**Comment by TyOverby** [comment]
@dgriffen: I also just published `1.0.0-alpha1` which includes your serde 0.9.* changes *and* my rustc-serialize removal.
